From 9f67c915cd6710ef636bb9c28420b7fe7576c148 Mon Sep 17 00:00:00 2001
From: Sam Day <me@samcday.com>
Date: Sun, 4 Oct 2020 22:09:47 +0200
Subject: [PATCH] [qmemman] systemd improvements

 * log straight to journal with official systemd JournalHandler
 * add socket activation support to daemon, along with socket unit file
 * use systemd.daemon.notify() rather than manual code writing to socket
---
 linux/systemd/Makefile              |  1 +
 linux/systemd/qubes-qmemman.service |  1 -
 linux/systemd/qubes-qmemman.socket  |  6 +++
 qubes/tools/qmemmand.py             | 79 ++++++++++++++++-------------
 4 files changed, 51 insertions(+), 36 deletions(-)
 create mode 100644 linux/systemd/qubes-qmemman.socket

diff --git a/linux/systemd/Makefile b/linux/systemd/Makefile
index e0cbbf60..fc7bdd77 100644
--- a/linux/systemd/Makefile
+++ b/linux/systemd/Makefile
@@ -8,6 +8,7 @@ install:
 	cp qubes-core.service $(DESTDIR)$(UNITDIR)
 	cp qubes-vm@.service $(DESTDIR)$(UNITDIR)
 	cp qubes-qmemman.service $(DESTDIR)$(UNITDIR)
+	cp qubes-qmemman.socket $(DESTDIR)$(UNITDIR)
 	cp qubesd.service $(DESTDIR)$(UNITDIR)
 	install -d $(DESTDIR)$(UNITDIR)/lvm2-pvscan@.service.d
 	install -m 0644 lvm2-pvscan@.service.d_30_qubes.conf \
diff --git a/linux/systemd/qubes-qmemman.service b/linux/systemd/qubes-qmemman.service
index a2830ef0..ac4caffc 100644
--- a/linux/systemd/qubes-qmemman.service
+++ b/linux/systemd/qubes-qmemman.service
@@ -4,7 +4,6 @@ Description=Qubes memory management daemon
 [Service]
 Type=notify
 ExecStart=/usr/bin/qmemmand
-StandardOutput=syslog
 
 [Install]
 WantedBy=multi-user.target
diff --git a/linux/systemd/qubes-qmemman.socket b/linux/systemd/qubes-qmemman.socket
new file mode 100644
index 00000000..023187be
--- /dev/null
+++ b/linux/systemd/qubes-qmemman.socket
@@ -0,0 +1,6 @@
+[Socket]
+ListenStream=/run/qubes/qmemman.sock
+SocketMode=0600
+
+[Install]
+WantedBy=sockets.target
diff --git a/qubes/tools/qmemmand.py b/qubes/tools/qmemmand.py
index 1cb23eb4..ffc5eed2 100644
--- a/qubes/tools/qmemmand.py
+++ b/qubes/tools/qmemmand.py
@@ -28,6 +28,8 @@ import socket
 import sys
 import threading
 
+import systemd.daemon
+import systemd.journal
 import xen.lowlevel.xs
 
 import qubes.qmemman
@@ -235,27 +237,33 @@ parser.add_argument('--foreground',
 
 
 def main():
-    args = parser.parse_args()
+    under_systemd = 'NOTIFY_SOCKET' in os.environ
 
-    # setup logging
-    ha_syslog = logging.handlers.SysLogHandler('/dev/log')
-    ha_syslog.setFormatter(
-        logging.Formatter('%(name)s[%(process)d]: %(message)s'))
-    logging.root.addHandler(ha_syslog)
+    args = parser.parse_args()
 
-    # leave log for backwards compatibility
-    ha_file = logging.FileHandler(LOG_PATH)
-    ha_file.setFormatter(
-        logging.Formatter('%(asctime)s %(name)s[%(process)d]: %(message)s'))
-    logging.root.addHandler(ha_file)
+    if not under_systemd:
+        # setup logging
+        ha_syslog = logging.handlers.SysLogHandler('/dev/log')
+        ha_syslog.setFormatter(
+            logging.Formatter('%(name)s[%(process)d]: %(message)s'))
+        logging.root.addHandler(ha_syslog)
 
-    if args.foreground:
-        ha_stderr = logging.StreamHandler(sys.stderr)
+        # leave log for backwards compatibility
+        ha_file = logging.FileHandler(LOG_PATH)
         ha_file.setFormatter(
             logging.Formatter('%(asctime)s %(name)s[%(process)d]: %(message)s'))
-        logging.root.addHandler(ha_stderr)
-
-    sys.stdin.close()
+        logging.root.addHandler(ha_file)
+
+        if args.foreground:
+            ha_stderr = logging.StreamHandler(sys.stderr)
+            ha_file.setFormatter(
+                logging.Formatter('%(asctime)s %(name)s[%(process)d]: %(message)s'))
+            logging.root.addHandler(ha_stderr)
+        sys.stdin.close()
+    else:
+        # Under systemd we use the systemd JournalHandler.
+        logging.root.addHandler(
+            systemd.journal.JournalHandler(SYSLOG_IDENTIFIER="qmemman"))
 
     logging.root.setLevel(parser.get_loglevel_from_verbosity(args))
 
@@ -281,30 +289,31 @@ def main():
         ' CACHE_FACTOR={algo.CACHE_FACTOR}'.format(
             algo=qubes.qmemman.algo))
 
-    try:
-        os.unlink(SOCK_PATH)
-    except:
-        pass
+    server = None
+    systemd_sock = None
+    systemd_listen_fds = systemd.daemon.listen_fds(True)
+    if systemd_listen_fds:
+        log.debug('listening on systemd provided socket')
+        server = socketserver.UnixStreamServer(None, QMemmanReqHandler, False)
+        server.socket = socket.fromfd(
+            systemd_listen_fds[0], socket.AF_UNIX, socket.SOCK_STREAM)
+    else:
+        try:
+            os.unlink(SOCK_PATH)
+        except:
+            pass
+
+        log.debug('instantiating server')
+        os.umask(0)
 
-    log.debug('instantiating server')
-    os.umask(0)
+        server = socketserver.UnixStreamServer(SOCK_PATH, QMemmanReqHandler)
+        os.umask(0o077)
 
     # Initialize the connection to Xen and to XenStore
     system_state.init()
 
-    server = socketserver.UnixStreamServer(SOCK_PATH, QMemmanReqHandler)
-    os.umask(0o077)
-
-    # notify systemd
-    nofity_socket = os.getenv('NOTIFY_SOCKET')
-    if nofity_socket:
-        log.debug('notifying systemd')
-        s = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
-        if nofity_socket.startswith('@'):
-            nofity_socket = '\0%s' % nofity_socket[1:]
-        s.connect(nofity_socket)
-        s.sendall(b"READY=1")
-        s.close()
+    if under_systemd:
+        systemd.daemon.notify('READY=1')
 
     threading.Thread(target=server.serve_forever).start()
     XS_Watcher().watch_loop()
-- 
2.28.0

